### ISSUE TITLE
[Feature] Enable Custom Headers to be sent in DocumentDB Requests

### ISSUE THREAD

**Issue by PlagueHO** [Azure/azure-documentdb-dotnet#412]
We would like to put a reverse proxy between calls from our (Azure Function) APIs and the CosmosDB endpoint to enable us to implement a second layer of security validation around the CosmosDB endpoint. We would like to be able to include a JWT as an additional request header in all calls to the CosmosDB.

The reverse proxy would receive the request, validate the JWT in the additional header (stripping this header at the same time) and then forwarding the request (without additional header) onto the CosmosDB. Optionally we could implement client certificate mutual auth with the proxy, but this may be a bigger change.

This would allow us to satisfy the requirements of our security team while still using the DocumentDB SDK and Azure Function consumption plans.

Is there a plan to OSS this SDK as then we could then add this feature ourselves. If not, is this something that is on the roadmap or even possible?

**Comment by kirankumarkolli** [comment]
@PlagueHO this is an abstraction which can be achieved on top of existing SDK. Please check sample at https://github.com/Azure/azure-documentdb-dotnet/tree/master/samples/xamarin also covers similar scenario.

**Comment by PlagueHO** [comment]
Hi @kirankumarkolli - first of all, thank you for your response! It is much appreciated.

I am aware of this pattern and had actually implemented as a POC. Unfortunately our security team is still not happy with the approach because it does not reduce the **impact** of a CosmosDB master key being obtained by an attacker. They do agree that with proper management of the master key (Key Vault, RBAC restrictions on the CosmosDB and Key Vault etc) the likelihood of breach is extremely low.

If we are able to implement a transparent reverse proxy in front of the CosmosDB we would be able to use the CosmosDB IP firewall to only accept HTTPS requests from the reverse proxy - which could provide pre-authentication by way of mutual client certificate authentication and/or JWT validation.

We can easily implement the reverse proxy (using something like Kestrel), but the DocumentDB SDK as it currently stands does not allow us to include any custom headers or provide the mutual client certificate authentication. If this DocumentDB SDK can not be extended to provide this then we'll have to roll our own DocumentDB SDK - which is not hugely difficult, but why reinvent the wheel if we don't have to? :grin:

tl;dr;
The Token Broker approach (what you linked to) is not acceptable by our security team because the CosmosDB front end does not enable multiple layers of security (HMAC key only - IP firewall will not work for consumption plan functions because outbound IP not known/static).

**Comment by kirankumarkolli** [comment]
@PlagueHO does limiting access to cosmos account to an azure VNET work in this scenario? 

With that cosmosdb will only accept requriest for you configured account from with-in that VNET.
